api/python/quilt3/backends/base.py

```diff
@@ -0,0 +1,174 @@
+
+    def manifests_package_dir(self, pkg_name) -> PhysicalKey:
+        return self.manifests_global_dir  # TODO: does this make sense?
+
+    def manifest_pk(self, pkg_name: str, top_hash: str) -> PhysicalKey:
+        return self.root.join(f'packages/{top_hash}')
+
+    def delete_package(self, pkg_name: str, top_hash: str = None):
+        package_path = self.pointers_dir(pkg_name)
+        paths = list(list_url(package_path))
+        if not paths:
+            raise QuiltException("No such package exists in the given directory.")
+
+        if top_hash is not None:
+            top_hash = self.resolve_top_hash(pkg_name, top_hash)
+            deleted = []
+            remaining = []
+            for path, pkg_hash in self.list_package_versions(pkg_name):
+                (deleted if pkg_hash == top_hash else remaining).append(path)
+            if not deleted:
+                raise QuiltException("No such package version exists in the given directory.")
+            for path in deleted:
+                delete_url(package_path.join(path))
+            if 'latest' in deleted and remaining:
+                # Create a new "latest". Technically, we need to compare numerically,
+                # but string comparisons will be fine till year 2286.
+                new_latest = max(remaining)
+                copy_file(package_path.join(new_latest), package_path.join('latest'))
+        else:
+            for path, _ in paths:
+                delete_url(package_path.join(path))
+
+    def push_manifest(self, pkg_name: str, top_hash: str, manifest_data: bytes):
+        put_bytes(manifest_data, self.manifest_pk(pkg_name, top_hash))
+        hash_bytes = top_hash.encode()
+        named_path = self.pointers_dir(pkg_name)
+        # TODO: use a float to string formatter instead of double casting
+        timestamp_path = named_path.join(str(int(time.time())))
+        put_bytes(hash_bytes, timestamp_path)
+        put_bytes(hash_bytes, self.pointer_latest_pk(pkg_name))
+
+    def resolve_top_hash(self, pkg_name: str, hash_prefix: str) -> str:
+        if len(hash_prefix) == 64:
+            top_hash = hash_prefix
+        elif 6 <= len(hash_prefix) < 64:
+            matching_hashes = [h for h, _
+                               in list_url(self.manifests_package_dir(pkg_name))
+                               if h.startswith(hash_prefix)]
+            if not matching_hashes:
+                raise QuiltException("Found zero matches for %r" % hash_prefix)
+            elif len(matching_hashes) > 1:
+                raise QuiltException("Found multiple matches: %r" % hash_prefix)
+            else:
+                top_hash = matching_hashes[0]
+        else:
+            raise QuiltException("Invalid hash: %r" % hash_prefix)
+        # TODO: verify that name is correct with respect to this top_hash
+        return top_hash
+
+    def shorten_top_hash(self, pkg_name: str, top_hash: str) -> str:
+        min_shorthash_len = 7
```

Comment:
a few lines above the minimum is 6; i guess that's fine if you're supporting legacy short hashes of 6 chars, but maybe a comment and why not a constant for this: `MIN_SHORTHASH_LEN`?

Reply:
This is just a copy of [the current implementation](https://github.com/quiltdata/quilt/blob/ad3b381657f07befc818930d83cae2766300fc2d/api/python/quilt3/packages.py#L480-L519). If we want to change something, we should do it in `master`, but not here.